### ISSUE TITLE
use of cookies only if user accepts cookies consent

### ISSUE THREAD

**Issue by javierguzman** [i18next/react-i18next#1362]
Hello all,

Sorry to put this here but I have been asking around without any response and also I am not sure if this is the expected behavior.

I have been trying to tell react-i18next to use cookies only after user accepts them. My initial thought has been to init i18next without cookies and once they are accepted, to init again with cookies enabled. My first question is, would this create side effects? For example, if you init twice, will the user lose the language selected previously?

Also, I have tried to implement this mechanism I have mentioned but the cookie still appears before accepting. Here is the code (right now only with English enabled):

```
import i18n from 'i18next';
import HttpApi from 'i18next-http-backend';
import LanguageDetector from 'i18next-browser-languagedetector';
import { initReactI18next } from 'react-i18next';
const i18nInitialConfig = (): void => {
  i18n
    .use(HttpApi)
    .use(LanguageDetector)
    .use(initReactI18next)
    // init i18next
    // for all options read: https://www.i18next.com/overview/configuration-options
    .init({
      fallbackLng: 'en',
      whitelist: ['en'],
      preload: ['en'],
      load: 'languageOnly',
      ns: namespaces,
      debug: false,
      interpolation: {
        escapeValue: false // not needed for react as it escapes by default
      },
      detection: {
        order: ['querystring'],
        lookupQuerystring: 'lng'
      },
      react: {
        // useSuspense: false,
      }
    });
};
```

The idea is that in detection I only have querystring at the very beginning.  Once the user accepts the cookies, I init with cookies options added into the detection field. However, even before the second init I can see the cookies in my browser. Is this the expected behavior?

I would appreciate any help.
Thank you in advance and regards.

**Comment by adrai** [comment]
i18next-browser-language-detector is using cookies: https://github.com/i18next/i18next-browser-languageDetector
If you do not like to use cookies, set the caches option accordingly: https://github.com/i18next/i18next-browser-languageDetector#detector-options
i.e. https://github.com/i18next/i18next-browser-languageDetector/issues/233#issuecomment-773943376

**Comment by javierguzman** [comment]
Thanks for the quick response @adrai !! I am going to give it a try and if everything works I will close the issue. By the way, what about the side effects I asked about this mechanism? I mean, is it ok to init twice? Or do you expect any problems with that? Just asking because I have played around with some init functions in the past that would break stuff if called twice. Thank you again.

**Comment by adrai** [comment]
regarding the init... there will be at least a warning... maybe use createInstance or cloneInstance

https://www.i18next.com/overview/api#instance-creation

**Comment by javierguzman** [comment]
Thanks again @adrai I have tried with caches: [] but the cookie is still set. Also, if I use createInstance and then I call it later, how can I remove the previous instance?

**Comment by adrai** [comment]
Just tried it with this example, and it worked: https://codesandbox.io/s/react-i18next-example-forked-qricr?file=/src/i18n.js

**Comment by adrai** [comment]
regarding the instances, what do you exactly mean with "remove" the previous instance? - just pass the new instance to your provider: https://react.i18next.com/latest/i18nextprovider
btw: if you don't want to deal with different i18next instances, you may need to refresh the page....

**Comment by javierguzman** [comment]
Hello @adrai, thanks as always; regarding the i18next instances I was not using before the i18nextprovider and based on what I have read it makes sense as it seems to be used only on SSR or if I need to manage multiple instances (which is my case now). I have managed to implement (I think successfully) the handle of multiple instances. I have created the following GIST with the component skeleton just in case it can help someone in future. Briefly, I just pass to the provider the current configuration instance and render the rest of the app within:

https://gist.github.com/javierguzman/ef7ef7cfa10d0855555901d1bd4e6069

On top of that, I have discovered the problem with caches: [] "not working". Indeed that makes the cookie not to be set, however, I did not realize my backend (nodejs) uses i18next as well and it was the one setting the cookie. Therefore, my final question would be, if I leave this in nodejs:
```
  detection: {
    order: ['querystring', 'cookie', 'header'],
    caches: [],
    lookupCookie: 'lng'
  },
```
  
Would that mean that cookie is not set but still it will look for the cookie named lng? Just asking because if the answer is yes, it would be perfect as it would be working ok, otherwise I would need to implement a mechanism to let nodejs the user has accepted cookies, etc.

**Comment by adrai** [comment]
> Would that mean that cookie is not set but still it will look for the cookie named lng? Just asking because if the answer is yes, it would be perfect as it would be working ok, otherwise I would need to implement a mechanism to let nodejs the user has accepted cookies, etc.

you mean with [i18next-http-middleware](https://github.com/i18next/i18next-http-middleware#language-detection) ? -> yes that should work

**Comment by javierguzman** [comment]
Yes, with i18next-http-middleware. Perfect then! I will close then this issue and if I come with some error in future I will open another one. I am still hesitant about changing my backend as I really need to test this again very thoroughly, I am worried there is a scenario where the front-end uses X language and my back-end believes the user uses Y language. But so far with the quick check seems to be ok. Thank you very much for your help @adrai :)

**Comment by adrai** [comment]
If you like this module don’t forget to star this repo. Make a tweet, share the word or have a look at our https://locize.com to support the devs of this project.

There are many ways to help this project 🙏
